### ISSUE TITLE
Different Augur versions between amd64 and arm64 images

### ISSUE THREAD

**Issue by victorlin** [nextstrain/docker-base#128]
### Current Behavior

It is possible that images of the same tag for different architectures have different Augur versions.

### Expected behavior

Images of the same tag for different architectures should be guaranteed to have the same Augur version.

### How to reproduce

Not sure that this is easily reproducible, but it was brought up by @corneliusroemer [in Slack](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1676906669414509) for the Augur 21.0.0 → 21.0.1 update.

In the [triggered image rebuild](https://github.com/nextstrain/docker-base/actions/runs/4206926139/jobs/7301049323), the two images pulled different versions of Augur:

```
2023-02-17T19:29:48.6526581Z #90 [linux/amd64 builder 42/47] RUN /builder-scripts/download-repo https://github.com/nextstrain/augur "$(/builder-scripts/latest-augur-release-tag)" .  && pip3 install --editable .
… nextstrain-augur==21.0.0

2023-02-17T19:30:52.5529403Z #99 [linux/arm64 builder 42/47] RUN /builder-scripts/download-repo https://github.com/nextstrain/augur "$(/builder-scripts/latest-augur-release-tag)" .  && pip3 install --editable .
… nextstrain-augur==21.0.1
```

A [relevant line](https://github.com/nextstrain/augur/actions/runs/4206823143/jobs/7300812344#step:10:54) from the Augur 21.0.1 release job logs shows when the new version was published to PyPI:

```
2023-02-17T19:28:11.7789440Z https://pypi.org/project/nextstrain-augur/21.0.1/
```

My speculation is that [the PyPI endpoint used to retrieve the version](https://github.com/nextstrain/docker-base/blob/d59452c4800c09e7f418c80d70379523a7cf84ef/builder-scripts/latest-augur-release-tag#LL11C12-L11C55) was only updated sometime within the 1 minute difference between the two invocations of `latest-augur-release-tag`.

### Possible solution

1. Determine the latest Augur version using the latest GitHub Release or by parsing `__version__.py`, instead of a PyPI endpoint that has some lag time.
2. Only trigger the Docker image rebuild after the new Augur version is known to be available at the PyPI endpoint used to retrieve the version.
3. Add tests to verify that Augur versions (and other software versions) are the same before publishing the public final images.
4. Catch this sort of thing programmatically by testing more than just the zika workflow. I think we can pretty easily run the existing CI workflows using [pathogen-repo-ci](https://github.com/nextstrain/.github/blob/24efb1981ec34b188535da9b1307076a06c1f3f3/.github/workflows/pathogen-repo-ci.yaml) ([monkeypox](https://github.com/nextstrain/monkeypox/blob/b54768ec17872eb0d898e29527785642f6b98c0d/.github/workflows/ci.yaml#L8-L9), [ncov](https://github.com/nextstrain/ncov/blob/b2095c0e502441ebcb0a3054418c9954cf897439/.github/workflows/ci.yaml#L8-L11)) with a bit of tinkering to allow specifying the Docker image used for the runtime.

**Comment by corneliusroemer** [comment]
Agree with your analysis, that's also what I figured out. Though the logs show `[linux/arm64 builder]` twice, rather than the required `amd` once and `arm` the other time.

**Comment by huddlej** [comment]
> We should just not get the release version from Pypi, but use the release as tagged in Github releases in the Augur repo directly! Or use the latest tag.

Yeah, if we're pulling the code from GitHub, we might as well use the latest version on GitHub. The latest tag is probably too fragile, since we can add non-version tags at any time that could be technically the latest. [We store the version info in Augur's code itself](https://github.com/nextstrain/augur/blob/f4e46fba552297577dcbbc0bd3644639452a1591/augur/__version__.py#L1), which might be the most reliable way to find the latest version tag. Or maybe we could pass the version information from the Augur release CI job to the docker-base rebuild CI job? Not sure if GH Action workflows accept arbitrary command line arguments...

**Comment by corneliusroemer** [comment]
> Yeah, if we're pulling the code from GitHub, we might as well use the latest version on GitHub.

Exactly! Why not query for the latest Github release? Since we now make these in an automated fashion that should work?

**Comment by huddlej** [comment]
Do you mean [the latest GitHub "Release"](https://github.com/nextstrain/augur/releases)? Those aren't automated; we still manually copy/paste content to create them. Although, automating those and then pulling the latest release would be great!

**Comment by corneliusroemer** [comment]
Oh I didn't know this wasn't automated. Then my proposal doesn't work, although automation of the GitHub release should not be too hard, we could easily create some release and edit the notes later.

**Comment by joverlee521** [comment]
> We should just not get the release version from Pypi, but use the release as tagged in Github releases in the Augur repo directly! Or use the latest tag.

Hmm, changing the release version source to GitHub wouldn't completely solve the issue. If someone created a new release on GitHub while the job is running then we could end up with the same issue. 

I think the most stable path would be to add checks to verify the versions are the same between the two images. Either error if they differ for us to intervene or automatically re-run the build that has the earlier version.

**Comment by corneliusroemer** [comment]
> If someone created a new release on GitHub while the job is running then we could end up with the same issue.

But why would that happen? Why would someone create a new release minutes after a previous one? Even then, this would trigger another docker build which would use the latest release - so only the intermediate docker build (not the latest) would be at risk of having different versions.

A simple delay should fix it in almost all situations, using tags would be possible, just mean we can't use tags for anything else - though we've never used tags for anything but versions. Checking versions are the same between the images (not just augur but also other packages) is a bonus.

<img width="1324" alt="image" src="https://user-images.githubusercontent.com/25161793/220479535-f48a37b4-772a-4500-8b46-7eaaaf43a5a5.png">

**Comment by joverlee521** [comment]
> But why would that happen? Why would someone create a new release minutes after a previous one? 

I'm not saying it would happen, just that if it does happen we will see the same issue again. 

> Even then, this would trigger another docker build which would use the latest release - so only the intermediate docker build (not the latest) would be at risk of having different versions.

True, but I thought we want to avoid the two images differing at all, not just in the latest. 
Ah, actually should we delete the `nextstrain/base:build-20230217T192833Z` images from docker hub or somehow label them as differing?

**Comment by victorlin** [comment]
@corneliusroemer

> Though the logs show [linux/arm64 builder] twice, rather than the required amd once and arm the other time.

Oops good catch! I'll update the issue description.

> although automation of the GitHub release should not be too hard

I just picked up on an old PR which implements this: https://github.com/nextstrain/augur/pull/957

**Comment by victorlin** [comment]
@huddlej

> if we're pulling the code from GitHub, we might as well use the latest version on GitHub.

This is a good point that I didn't realize! I agree, there is no good reason to use the version from PyPI unless Augur is being downloaded from there.

Once GitHub Releases are automated, I imagine pulling from that would be easier than parsing `__version__.py`.

**Comment by victorlin** [comment]
@joverlee521

> True, but I thought we want to avoid the two images differing at all, not just in the latest.

Yes, this is still a good aim! 2 separate issues are being discussed here:

1. `devel/latest-augur-release-tag` unnecessarily sources the version from PyPI when Dockerfile does not use PyPI.
2. Build commands in the Dockerfile can produce different results at different times. Checks should be put in place to reduce the effects of this when building a multi-arch image (which is really 2 images built separately).

(1) can be addressed by changing the release version source, and (2) can be addressed by adding tests before publishing the public final images.

> Ah, actually should we delete the `nextstrain/base:build-20230217T192833Z` images from docker hub or somehow label them as differing?

I think it's best to leave it there, like how we don't "un-release" other software versions once tagged/released. Most use cases of the Docker image should automatically pull the latest, and it's recommended for others to manually pull the latest via `nextstrain update`.

**Comment by joverlee521** [comment]
> 1. devel/latest-augur-release-tag unnecessarily sources the version from PyPI when Dockerfile does not use PyPI.

Conversation from the original [PR](https://github.com/nextstrain/docker-base/pull/44) that added the PyPI release version source is relevant here. We should decide on if we want to stick with GitHub as the source for Augur in docker-base or move to pull Augur from PyPI to match the other runtimes.

**Comment by victorlin** [comment]
@joverlee521 thanks for pointing that out, I had forgotten. Created a separate issue #129, since the topic is tangential from the original issue here.

**Comment by corneliusroemer** [comment]
> pull Augur from PyPI to match the other runtimes

Our conda-base environment pulls from _bioconda_ 🙃  So there's already inconsistency anyways. Using Github means we don't rely on servers that may take variable amounts of time to show changes. 

What other runtimes do we have?

**Comment by victorlin** [comment]
> The Augur (and Auspice) release process that triggers an image build should pass the version it just released to the image build process.

At first I thought this sounded good, but I'm having second thoughts now.

It only makes an invariant for the specified software, not others. Let's say we want to implement this for both Augur and Auspice. It would look something like this:

```sh
# Upon Augur release
gh workflow run ci.yml --repo nextstrain/docker-base -f augur_ref=21.0.1

# Upon Auspice release
gh workflow run ci.yml --repo nextstrain/docker-base -f auspice_ref=2.45.0
```

When triggering from a new Augur release, there would still be a need to determine the latest Auspice version (still variant), and vice-versa.

Since the specified version info is only applicable to the current build, this feels like "half-pinning" to me (where full pinning would be like `pip3 install augur==21.0.1`). I'd rather:

1. Improve the way the latest version is fetched (only reduces failures, but they can be caught by checking across platforms), or
2. Implement full pinning somehow where the specified version info is checked into the repo to be used by all builds.

**Comment by tsibley** [comment]
> It only makes an invariant for the specified software, not others. … When triggering from a new Augur release, there would still be a need to determine the latest Auspice version (still variant).

Yes, but I think that'll be fine in almost all cases (as @corneliusroemer says too).

The invariant I'm suggesting we establish isn't on all versions, but on the just-released version that's the reason we're triggering the rebuild in the first place.

> Implement full pinning somehow where the specified version info is checked into the repo to be used by all builds.

I do agree that this is a more robust alternative, but it complicates a bit the triggering of new builds after a version bump. Maybe worth it! (But the easier/quicker thing seemed like it reduce failures enough.)

**Comment by tsibley** [comment]
We'll also want to consider that any solution we choose here we will also want to apply to conda-base (if not now, [in the future](https://github.com/nextstrain/conda-base/issues/2)).

**Comment by victorlin** [comment]
Closing this issue since the original issue (different Augur versions between image variants) has been addressed by #130.
